src/operators_lazytensor.jl

```diff
@@ -96,21 +96,45 @@ isequal(x::LazyTensor, y::LazyTensor) = samebases(x,y) && isequal(x.indices, y.i
 # Arithmetic operations
 -(a::LazyTensor) = LazyTensor(a, -a.factor)
 
-function +(a::LazyTensor{B1,B2}, b::LazyTensor{B1,B2}) where {B1,B2}
-    if length(a.indices) == 1 && a.indices == b.indices
+const single_dataoperator{B1,B2} = LazyTensor{B1,B2,F,I,Tuple{T}} where {B1,B2,F,I,T<:DataOperator} 
+function +(a::T1,b::T2) where {T1 <: single_dataoperator{B1,B2},T2 <: single_dataoperator{B1,B2}} where {B1,B2}
+    if a.indices == b.indices
         op = a.operators[1] * a.factor + b.operators[1] * b.factor
         return LazyTensor(a.basis_l, a.basis_r, a.indices, (op,))
     end
-    throw(ArgumentError("Addition of LazyTensor operators is only defined in case both operators act nontrivially on the same, single tensor factor."))
+    LazySum(a) + LazySum(b)
 end
-
-function -(a::LazyTensor{B1,B2}, b::LazyTensor{B1,B2}) where {B1,B2}
-    if length(a.indices) == 1 && a.indices == b.indices
+function -(a::T1,b::T2) where {T1 <: single_dataoperator{B1,B2},T2 <: single_dataoperator{B1,B2}} where {B1,B2}
+    if a.indices == b.indices
         op = a.operators[1] * a.factor - b.operators[1] * b.factor
         return LazyTensor(a.basis_l, a.basis_r, a.indices, (op,))
     end
-    throw(ArgumentError("Subtraction of LazyTensor operators is only defined in case both operators act nontrivially on the same, single tensor factor."))
+    LazySum(a) - LazySum(b)
+end
+
+function tensor(a::LazyTensor{B1,B2},b::AbstractOperator{B3,B4}) where {B1,B2,B3,B4}
+    if isequal(b,identityoperator(b))
+        btotal_l = a.basis_l ⊗ b.basis_l
+        btotal_r = a.basis_r ⊗ b.basis_r
+        LazyTensor(btotal_l,btotal_r,a.indices,(a.operators...,),a.factor)
+    elseif B3 <: CompositeBasis || B4 <: CompositeBasis
+        throw(ArgumentError("tensor(a::LazyTensor{B1,B2},b::AbstractOperator{B3,B4}) is not implemented for B3 or B4 being CompositeBasis unless b is identityoperator "))
+    else
+        a ⊗ LazyTensor(b.basis_l,b.basis_r,[1],(b,),1)
+    end
```

Comment:
How should we treat LazyTensor with non-equivalent LR and BR. The first if statement is only well defined for BL==BR it seems.

Reply:
Yeah, so LazyTensor with non-rectangular factors in the composite basis implicitly defines simple isometries, like you constructed in your new implementation of `identityoperator()`. Honestly I think it's okay to just not handle the `b is an identity` case specially at all and just include the explicit identity/isometry operator. It's kind of unfortunate to explicitly check that a matrix is the identity here anyway if I'm expecting a lazy operation.

I have an `embed_lazy()` that I use to achieve this kind of thing.

Perhaps a nicer solution is to introduce a proper LazyIdentity/LazyIsometry operator that gets materialized only when necessary. Then we could implement `tensor(a::LazyTensor, b::LazyIdentity)` using something like your first conditional branch here.

Reply:
@amilsted , could you put up a gist with `embed_lazy` or just an incomplete pull request?

This discussion and `embed_lazy` would probably have a bearing on this issue as well: https://github.com/qojulia/QuantumOptics.jl/issues/322

Reply:
https://github.com/qojulia/QuantumOpticsBase.jl/pull/88

Reply:
I'd also say we can ignore the special `isequal(b,identityoperator(b))`. It should (hopefully) be the exception to run into this, so the impact shouldn't be too large.

One alternative way of solving this would be doing something like #90, but that's just an idea. For now I'd say don't special handle it, remove the new `identityoperator` method above and we're good to go.

By the way, thanks for all the work @mabuni1998 and thanks for the reviews @amilsted @Krastanov. I appreciate it!

Reply:
You're welcome! I now pushed a new version with the above things fixed. 

I would like to say that originally I put the check for identityoperator there because, in my own code, I noticed performance drops when doing stuff like the following (WaveguideBasis is a custom basis with operators defined as functions, not as matrices, and they, therefore, have to be Lazy): 
```
bc = FockBasis(2)
bw = WaveguideBasis(2,times)
adw = tensor(create(bc),identityoperator(bc),identityoperator(bc),destroy(bw))
```
Because here, identityoperator(bc) is explicitly stored in the resulting LazyTensor (there is a custom method creating LazyTensor here). Probably this is better solved by allowing operators spanning multiple basises to be stored in LazyTensor. Right now, we explicitly check for this case and don't allow it. However, this would require extra work (and I don't know how much) in the mul! method so let's leave it. This problem (I think) would not be solved by implementing LazyIdentity since identityoperator(bc) would not return such abjoct.

Also, I'm aware that in the above, one could just use embed or explicitly define LazyTensor. I simply just wanted to make it easier for the user to have better performance. 
Furthermore, my use case is probably niece, since I'm only using Lazy operators because my operators are defined as functions, whereas in most cases, you use it because evaluating the tensor product is expensive. I can just define custom methods that take care of this.